### PR TITLE
Introduce framework::Client

### PR DESCRIPTION
**Pull request by cschuet**
Introduces a framework::Client class that makes it more convenient to call gRPC methods.

[RFC=0002](https://github.com/googlecartographer/rfcs/blob/master/text/0002-cloud-based-mapping-1.md)

### CODE REVIEW

File: cartographer_grpc/framework/client.h [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+      const typename RpcHandlerType::RequestType& request) {
+    CHECK_EQ(rpc_method_.method_type(),
+             grpc::internal::RpcMethod::SERVER_STREAMING);
+    client_reader_.reset(
+        grpc::internal::ClientReaderFactory<
+            typename RpcHandlerType::ResponseType>::Create(channel_.get(),
+                                                           rpc_method_,
+                                                           &client_context_,
+                                                           request
+
```
Comment (gaschler) [inline cartographer_grpc/framework/client.h]: remove blank line

File: cartographer_grpc/framework/server_test.cc [hunk 2]
```diff
@@ -175,54 +176,49 @@ TEST_F(ServerTest, ProcessRpcStreamTest) {
       cartographer::common::make_unique<MathServerContext>());
   server_->Start();
 
-  proto::GetSumResponse result;
-  std::unique_ptr<grpc::ClientWriter<proto::GetSumRequest>> writer(
-      stub_->GetSum(&client_context_, &result));
```
Comment (gaschler) [inline cartographer_grpc/framework/server_test.cc]: Is client_context_ still used?

File: cartographer_grpc/framework/server_test.cc [hunk 3]
```diff
@@ -245,28 +237,32 @@ TEST_F(ServerTest, WriteFromOtherThread) {
     CHECK(responder());
   });
 
-  grpc::Status status = stub_->GetEcho(&client_context_, request, &result);
+  Client<GetEchoHandler> client(client_channel_);
+  proto::GetEchoResponse result;
```
Comment (gaschler) [inline cartographer_grpc/framework/server_test.cc]: result seems unused.

File: cartographer_grpc/framework/client.h (hunk 1 above)
Comment (cschuet) [reply cartographer_grpc/framework/client.h]: Done.

File: cartographer_grpc/framework/server_test.cc (hunk 2 above)
Comment (cschuet) [reply cartographer_grpc/framework/server_test.cc]: It's not. Removed!

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 1.

**Pull request merged**